rust-runtime/aws-smithy-http-server-python/src/middleware/layer.rs

```diff
@@ -0,0 +1,248 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Tower layer implementation of Python middleware handling.
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use aws_smithy_http_server::{
+    body::{Body, BoxBody},
+    protocols::Protocol,
+};
+use futures::{ready, Future};
+use http::{Request, Response};
+use pin_project_lite::pin_project;
+use pyo3::PyResult;
+use pyo3_asyncio::TaskLocals;
+use tower::{Layer, Service};
+
+use crate::{error::PyException, middleware::PyFuture, PyMiddlewares};
+
+/// Tower [Layer] implementation of Python middleware handling.
+///
+/// Middleware stored in the `handlers` attribute will be executed, in order,
+/// inside an async Tower middleware.
+#[derive(Debug, Clone)]
+pub struct PyMiddlewareLayer {
+    handlers: PyMiddlewares,
+    protocol: Protocol,
+    locals: TaskLocals,
+}
+
+impl PyMiddlewareLayer {
+    pub fn new(
+        handlers: PyMiddlewares,
+        protocol: &str,
+        locals: TaskLocals,
+    ) -> PyResult<PyMiddlewareLayer> {
+        let protocol = match protocol {
+            "aws.protocols#restJson1" => Protocol::RestJson1,
+            "aws.protocols#restXml" => Protocol::RestXml,
+            "aws.protocols#awsjson10" => Protocol::AwsJson10,
+            "aws.protocols#awsjson11" => Protocol::AwsJson11,
+            _ => {
+                return Err(PyException::new_err(format!(
+                    "Protocol {protocol} is not supported"
+                )))
+            }
+        };
+        Ok(Self {
+            handlers,
+            protocol,
+            locals,
+        })
+    }
+}
+
+impl<S> Layer<S> for PyMiddlewareLayer {
+    type Service = PyMiddlewareService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        PyMiddlewareService::new(
+            inner,
+            self.handlers.clone(),
+            self.protocol,
+            self.locals.clone(),
+        )
+    }
+}
+
+// Tower [Service] wrapping the Python middleware [Layer].
+#[derive(Clone, Debug)]
+pub struct PyMiddlewareService<S> {
+    inner: S,
+    handlers: PyMiddlewares,
+    protocol: Protocol,
+    locals: TaskLocals,
+}
+
+impl<S> PyMiddlewareService<S> {
+    pub fn new(
+        inner: S,
+        handlers: PyMiddlewares,
+        protocol: Protocol,
+        locals: TaskLocals,
+    ) -> PyMiddlewareService<S> {
+        Self {
+            inner,
+            handlers,
+            protocol,
+            locals,
+        }
+    }
+}
+
+impl<S> Service<Request<Body>> for PyMiddlewareService<S>
+where
+    S: Service<Request<Body>, Response = Response<BoxBody>> + Clone,
+{
+    type Response = Response<BoxBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let inner = self.inner.clone();
```

Comment:
Note about making this clone less expensive (Arc?)

Reply:
Done